### PR TITLE
Empty default facility and instrument

### LINKED ISSUE #29956: Help > About: "Do not show..." tickbox ignored
**Original reporter:** Vain

### Expected behavior
First time set-up screen (a.k.a. Help > About mantidworkbench interface) has the `Do not show again until next version` on the bottom left. When ticked and Workbench is restarted, the first-time set-up window does not appear
### Actual behavior
If the facility is not set in the Mantid.user.properties file then **this tickbox is ignored** and the first time set-up window appears every time

### Steps to reproduce the behavior
Edit `Mantid.user.properties` file to set the `default.facility=` as blank!
Open workbench and on the first-time setup interface, tick this "Do not show..." box.
Hit close and also close workbench.
Reopen workbench and the firstime set-up window will reappear! 
### Platforms affected
Found on Windows Surface Pro Workbench 5.1.1

### PR DESCRIPTION
**Pull request by srikanthravipati**
**Description of work.**

A new empty facility with empty instrument is the default facility.

User has to select their choice of facility for the first time, including ISIS, and instrument.

Before, if the facility in Mantid.user.properties is empty (for whatever reason), in the GUI
it was inconsistently showing it to be ISIS.

Now, if the facility and instrument are indeed empty in Mantid.user.properties file, it will be consistently 
reflected in the GUI and user will have to select their choices.

**Report to:**  Vain

**To test:**

1. Set facility value in Mantid.user.properties file to be empty. (Re)Open workbench and check 
    if it is consistently reflected in the workbench GUI (that the facility and instrument are indeed empty)

2. Select required facility and instrument, and also check 'Do not show again until next version'

3. Close and reopen workbench. First time set-up window should not appear now. 

Refs #29956 

Fixes #29956 and close #29956 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocessTest.py [hunk 1]
```diff
@@ -119,6 +120,8 @@ def testDefaultRunFIGARO(self):
         self.assertAlmostEqual(numpy.rad2deg(two_theta_fg), 0., delta=0.1)
 
     def testTwoInputFiles(self):
+        ConfigService.Instance().setString("default.facility",   "ILL")
```
Comment (darominski) [inline Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocessTest.py:123]: Perhaps overriding setUp and expanding the already-present tearDown methods would be more elegant, the solution could look as follows:
```
    def setUp(self):
        self._facility = config['default.facility']
        self._instrument = config['default.instrument']

        config['default.facility'] = 'ILL'
        config['default.instrument'] = 'D17'

    def tearDown(self):
        if self._facility:
            config['default.facility'] = self._facility
        if self._instrument:
            config['default.instrument'] = self._instrument
        mtd.clear()
```

**Review by darominski — COMMENTED** [review]
I seem to be unable to reproduce the desired feature of not creating dialog even if blank was selected. 

First, I made sure the `default.facility` is set to blank and I was prompted with the welcome dialog, as expected. I set the blank facility and blank instrument, ticked 'Do not show...' checkbox, closed the dialog and the workbench. Upon restarting, I am prompted again with the same dialog. The dialog will not show up again only if a specific facility from the list but NOT blank is selected. Could you have a look at this?

File: Framework/DataHandling/test/LoadILLTest.h [hunk 2]
```diff
@@ -25,7 +25,17 @@ class LoadILLTest : public CxxTest::TestSuite {
   static LoadILLTest *createSuite() { return new LoadILLTest(); }
   static void destroySuite(LoadILLTest *suite) { delete suite; }
 
-  void tearDown() override { AnalysisDataService::Instance().clear(); }
+  void setUp() override {
+
+    ConfigService::Instance().setString("default.facility", "ISIS");
```
Comment (gvardany) [inline Framework/DataHandling/test/LoadILLTest.h:30]: Should this not be ILL?

Comment (srikanthravipati) [reply Framework/DataHandling/test/LoadILLTest.h:30]: yes, thanks!

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocessTest.py (hunk 1 above)
Comment (srikanthravipati) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocessTest.py]: Thanks for the suggestion. I have revised the section accordingly.

File: Framework/Kernel/src/InstrumentInfo.cpp [hunk 3]
```diff
@@ -39,9 +39,9 @@ InstrumentInfo::InstrumentInfo(const FacilityInfo *f,
     : m_facility(f) {
 
   m_name = elem->getAttribute("name");
-  if (m_name.empty()) {
-    throw std::runtime_error("Instrument name is not defined");
-  }
+  // if (m_name.empty()) {
```
[the remark this answers is not in the document]
Comment (srikanthravipati) [reply Framework/Kernel/src/InstrumentInfo.cpp]: Agreed and removed.

File: Framework/Kernel/src/FacilityInfo.cpp [hunk 4]
```diff
@@ -42,10 +42,10 @@ FacilityInfo::FacilityInfo(const Poco::XML::Element *elem)
       m_zeroPadding(0), m_delimiter(), m_extensions(), m_archiveSearch(),
       m_instruments(), m_noFilePrefix(), m_multiFileLimit(100),
       m_computeResources() {
-  if (m_name.empty()) {
-    g_log.error("Facility name is not defined");
-    throw std::runtime_error("Facility name is not defined");
-  }
+  // if (m_name.empty()) {
```
[the remark this answers is not in the document]
Comment (srikanthravipati) [reply Framework/Kernel/src/FacilityInfo.cpp]: Agreed and removed.

**Comment by srikanthravipati** [comment]
> I seem to be unable to reproduce the desired feature of not creating dialog even if blank was selected.
> 
> First, I made sure the `default.facility` is set to blank and I was prompted with the welcome dialog, as expected. I set the blank facility and blank instrument, ticked 'Do not show...' checkbox, closed the dialog and the workbench. Upon restarting, I am prompted again with the same dialog. The dialog will not show up again only if a specific facility from the list but NOT blank is selected. Could you have a look at this?

Thanks for taking time to review the PR.

In terms of testing, simply referring to #29956 is not very clear; sorry about that.
I have updated the description of work, and to test parts in the PR notes. 
Let me know if it's clear!

**Review by darominski — APPROVED** [review]
Thanks for the changes and updating the description. Now, I understand how it is supposed to work and it does work exactly like that right now. It's good to go after the RHEL tests pass.

**Pull request merged**